vmdb/app/models/ems_openstack_infra.rb

```diff
@@ -1,8 +1,24 @@
 class EmsOpenstackInfra < EmsInfra
   include EmsOpenstackMixin
 
+  before_create :ensure_parent_provider
```

Comment:
This might be better as `before_validation`.

Additionally you may need to set `:autosave => true` on the provider relation, but only in this subclass, because changes made in the ensure_parent_provider will not be saved without it